### PR TITLE
706 htaccess

### LINKED ISSUE #706: Use `.htaccess` files to secure configuration directories
Currently, the `config` directory is 'secured' by appending a hash string to the dir path. This has caused some issues when changing database credentials, etc., and could be improved.

I suggest instead adding a .htaccess file to the directory upon creation (or when there isn't one, see below) that denies access in the same way that the .htaccess file in the `private` files directory works.

Also, when you delete the .htaccess file from the files directory and re-submit the File System form (admin/config/media/file-system), it re-creates the .htaccess file automatically. The same should be true of the `config` directory, and maybe even add a field to the File System page where you can specify the `config` directory (rather than in 'settings.php').

Advocate: @BWPanda

---

~Original PR: https://github.com/backdrop/backdrop/pull/3238~
New PR to address feedback form code review https://github.com/backdrop/backdrop/pull/3279

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/706

Replaces https://github.com/backdrop/backdrop/pull/3238

### CODE REVIEW

**Comment by ghost** [comment]
@jenlampton Can you please refresh this? The deploy failed...

**Comment by jenlampton** [comment]
Closing for new sandbox.

**Comment by jenlampton** [comment]
New sandbox please

File: core/modules/config/tests/config.test [hunk 1]
```diff
@@ -158,6 +158,27 @@ class ConfigurationTest extends BackdropWebTestCase {
...
+    // Verify contents of .htaccess file.
+    $file = file_get_contents($active_dir . '/.htaccess');
+    $this->assertEqual($file, file_htaccess_lines(FALSE), 'The .htaccess file contains the proper content.');
+
+    // Remove .htaccess file to then test that it gets re-created.
+    @backdrop_unlink($active_dir . '/.htaccess');
+    $this->assertFalse(is_file($active_dir . '/.htaccess'), "Successfully removed the .htaccess file in the 'active' config directory.");
+
+    // Do a normal web request to confirm that the file is re-created.
+    $this->backdropGet('admin/config/reports/status');
```
Comment (indigoxela) [inline core/modules/config/tests/config.test]: That should be 'admin/reports/status'.

Additionally you need to create a user with permission to access that page.
```
    $admin_user = $this->backdropCreateUser(array(
      'access administration pages',
      'access site reports',
    ));
    $this->backdropLogin($admin_user);
    $this->backdropGet('admin/reports/status');
```

File: core/modules/config/tests/config.test [hunk 2]
```diff
@@ -158,6 +158,27 @@ class ConfigurationTest extends BackdropWebTestCase {
...
+   * Tests that config directories are protected by .htaccess files.
+   */
+  public function testConfigHtaccess() {
+    // Check that the .htacess file is present.
+    $active_dir = config_get_config_directory();
+    $this->assertTrue(is_file($active_dir . '/.htaccess'), "Successfully located the .htaccess file in the 'active' config directory.");
+
+    // Verify contents of .htaccess file.
+    $file = file_get_contents($active_dir . '/.htaccess');
+    $this->assertEqual($file, file_htaccess_lines(FALSE), 'The .htaccess file contains the proper content.');
```
Comment (indigoxela) [inline core/modules/config/tests/config.test]: file_htaccess_lines(FALSE) - would create content for public directories. You want private ones.

`file_htaccess_lines()`

Comment (jenlampton) [reply core/modules/config/tests/config.test]: thanks!

**Comment by quicksketch** [comment]
Merged into 1.x for 1.17.0.

**Pull request merged**